llama_index/chat_engine/types.py

```diff
@@ -195,6 +196,168 @@ async def aprint_response_stream(self) -> None:
             print(token, end="", flush=True)
 
 
+@dataclass
+class CohereStreamingAgentChatResponse(StreamingAgentChatResponse):
+    """Streaming chat response to user and writing to chat history."""
+    citations: List[dict] = field(default_factory=list)
+    documents: List[dict] = field(default_factory=list)
+    _documents_queue: queue.Queue = field(default_factory=queue.Queue)
+    _documents_aqueue: asyncio.Queue = field(default_factory=asyncio.Queue)
+    _citations_queue: queue.Queue = field(default_factory=queue.Queue)
+    _citations_aqueue: asyncio.Queue = field(default_factory=asyncio.Queue)
+
+    def put_in_queue(self, queue: queue.Queue, delta: Optional[str]) -> None:
+        queue.put_nowait(delta)
+
+    def aput_in_queue(self, queue: asyncio.Queue, delta: Optional[str]) -> None:
+        queue.put_nowait(delta)
+        self._new_item_event.set()
+
+    def write_response_to_history(
+        self, memory: BaseMemory, raise_error: bool = False
+    ) -> None:
+        if self.chat_stream is None:
+            raise ValueError(
+                "chat_stream is None. Cannot write to history without chat_stream."
+            )
+
+        # try/except to prevent hanging on error
+        try:
+            final_text = ""
+            for chat in self.chat_stream:
+                # Cohere response queue
+                self.put_in_queue(self._queue, chat.delta)
+                # Citations queue
+                if chat.raw.get("event_type", "") == StreamEvent.CITATION_GENERATION:
+                    self.put_in_queue(
+                        self._citations_queue, chat.raw.get("citations", [])
+                    )
+                # Documents queue
+                if chat.raw.get("event_type", "") == StreamEvent.SEARCH_RESULTS:
+                    self.put_in_queue(
+                        self._documents_queue, chat.raw.get("documents", [])
+                    )
+                # TODO  Ask TJ - maybe we'll need it
+                final_text += chat.delta or ""
```

Comment:
Do we have a `chat.delta` for all stream events, or only for the TEXT_GENERATION event?

Reply:
No, we have a chat.delta only for the TEXT_GENERATION event